debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/RetriableConnection.java

```diff
@@ -0,0 +1,219 @@
+        return executeWithRetry(func, null, name, rollback);
+    }
+
+    private synchronized <T> T executeWithRetry(ConnectionFunction<T> func, ConnectionFunctionVoid funcVoid, String name, boolean rollback)
+            throws SQLException, UncheckedIOException {
+        int attempt = 1;
+        while (true) {
+            if (conn == null) {
+                LOGGER.debug("Trying to reconnect (attempt {}).", attempt);
+                try {
+                    createConnection();
+                }
+                catch (SQLException e) {
+                    LOGGER.error("SQL Exception while trying to reconnect: " + e.getMessage(), e);
+                    conn = null;
+                    if (attempt >= maxRetryCount) {
+                        throw e;
+                    }
+                    attempt++;
+                    LOGGER.debug("Waiting for reconnect for {} ms.", waitRetryDelay);
+                    DelayStrategy delayStrategy = DelayStrategy.constant(waitRetryDelay);
+                    delayStrategy.sleepWhen(true);
+                    continue;
+                }
+            }
+            try {
+                if (func == null) {
+                    return func.accept(conn);
+                }
+                if (funcVoid == null) {
+                    funcVoid.accept(conn);
+                    return null;
+                }
+            }
+            catch (SQLRecoverableException e) {
+                LOGGER.warn("Attempt {} to call '{}' failed.", attempt, name, e);
+                if (rollback) {
+                    LOGGER.warn("'{}': doing rollback.", name);
+                    try {
+                        conn.rollback();
+                    }
+                    catch (SQLException ex) {
+                        // ignore
+                    }
+                }
+                try {
+                    conn.close();
+                }
+                catch (Exception ex) {
+                    // ignore
+                }
+                conn = null;
+
+            }
+            catch (SQLException e) {
+                LOGGER.warn("Call '{}' failed.", name, e);
+                if (rollback) {
+                    LOGGER.warn("'{}': doing rollback.", name);
+                    try {
+                        conn.rollback();
+                    }
+                    catch (SQLException ex) {
+                        // ignore
+                    }
+                }
+                throw e;
+            }
+        }
+    }
+
+    @FunctionalInterface
+    public interface ConnectionFunction<T> {
+        /**
+         * Performs this operation on the given connection.
+         *
+         * @param conn the input connection
+         * @return result of the operation
+         */
+        T accept(Connection conn) throws SQLException, UncheckedIOException;
+    }
+
+    @FunctionalInterface
+    public interface ConnectionFunctionVoid {
```

Comment:
Please don't call it `ConnectionFunction`. In Java it corresponds to `java.util.function.Consumer<T>` so it should be `ConnectionConsumer`.